Fixed #31507 -- Added QuerySet.exists() optimizations to compound queries.

David-Wobrock opened the pull request:
Ticket: https://code.djangoproject.com/ticket/31507

Hi,

I added the optimizations that we have on the `.exists()` call (removing selected columns for a "SELECT 1", removing ordering and limiting the number of results to 1) to a call with the `.union()` combinator. So we also strip the selected columns and remove ordering, but we [cannot add a limit clause in the subqueries](https://github.com/django/django/blob/master/django/db/models/sql/compiler.py#L442).

We don't apply the optimization for other combination predicates (difference, intersection) - should I add test cases for those, asserting that don't have this behaviour for these cases?

Cheers,
David

charettes commented on django/db/models/sql/query.py:
```diff
@@ -534,8 +534,13 @@ def exists(self):
             q.clear_select_clause()
+        if q.combined_queries and q.combinator == 'union':
+            q.combined_queries = tuple(
+                combined_query.exists(limit=False) for combined_query in q.combined_queries
```
I wonder if we want a database feature here for which backends support limiting within a combined query could be useful.

charettes commented on tests/queries/tests.py:
```diff
@@ -1937,6 +1937,18 @@ def test_exists(self):
         self.assertNotIn(id, qstr)
         self.assertNotIn(name, qstr)
 
+    def test_union_exists(self):
```
Move to `test_qs_combinators.py`

charettes reviewed (commented):
yeah I think additional tests for exists `intersect` and `difference` would be great.

David-Wobrock commented:
Hi @charettes, thanks for the review, once again 😄 

Splitted in 3 commits, so they can be amended/dropped more easily:
1. I moved the test into `test_qs_combinators.py` and added test cases for `.intersection()` and `.difference()` that fail if we change we include them in the optimization
2. I tried implementing a DB feature called `supports_limiting_in_compound`
3. adding the `.exists()` in the documentation examples of available functions for the combined `QuerySets`

charettes commented on django/db/models/sql/query.py:
```diff
@@ -534,8 +534,14 @@ def exists(self):
                 q.set_group_by(allow_aliases=False)
             q.clear_select_clause()
+        if q.combined_queries and q.combinator == 'union':
+            limit_combined = connections[DEFAULT_DB_ALIAS].features.supports_limiting_in_compound
```
I think using `connections[DEFAULT_DB_ALIAS]` is acceptable here as we even do it at a higher level (e.g. feature detection in `django.db.models`) but if we truly wanted to perform the check against a database the following minimal changes would allow you to use the proper database alias

```diff
diff --git a/django/db/models/expressions.py b/django/db/models/expressions.py
index c596ccfe5d..e356a5f981 100644
--- a/django/db/models/expressions.py
+++ b/django/db/models/expressions.py
@@ -1116,10 +1116,10 @@ def copy(self):
     def external_aliases(self):
         return self.query.external_aliases

-    def as_sql(self, compiler, connection, template=None, **extra_context):
+    def as_sql(self, compiler, connection, query=None, template=None, **extra_context):
         connection.ops.check_expression_support(self)
         template_params = {**self.extra, **extra_context}
-        subquery_sql, sql_params = self.query.as_sql(compiler, connection)
+        subquery_sql, sql_params = (query or self.query).as_sql(compiler, connection)
         template_params['subquery'] = subquery_sql[1:-1]

         template = template or template_params.get('template', self.template)
@@ -1142,7 +1142,6 @@ class Exists(Subquery):
     def __init__(self, queryset, negated=False, **kwargs):
         self.negated = negated
         super().__init__(queryset, **kwargs)
-        self.query = self.query.exists()

     def __invert__(self):
         clone = self.copy()
@@ -1150,7 +1149,9 @@ def __invert__(self):
         return clone

     def as_sql(self, compiler, connection, template=None, **extra_context):
-        sql, params = super().as_sql(compiler, connection, template, **extra_context)
+        sql, params = super().as_sql(
+            compiler, connection, template, query=self.query.exists(connection.alias), **extra_context
+        )
         if self.negated:
             sql = 'NOT {}'.format(sql)
         return sql, params
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index a7dadf5a40..0d29aa28f3 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -525,7 +525,7 @@ def get_count(self, using):
     def has_filters(self):
         return self.where

-    def exists(self):
+    def exists(self, using):
         q = self.clone()
         if not q.distinct:
             if q.group_by is True:
@@ -541,7 +541,7 @@ def exists(self):
         return q

     def has_results(self, using):
-        q = self.exists()
+        q = self.exists(using)
         compiler = q.get_compiler(using=using)
         return compiler.has_results()
```

charettes commented on tests/queries/test_qs_combinators.py:
```diff
@@ -232,6 +233,60 @@ def test_count_intersection(self):
         qs2 = Number.objects.filter(num__lte=5)
         self.assertEqual(qs1.intersection(qs2).count(), 1)
 
+    @skipIfDBFeature('supports_limiting_in_compound')
+    def test_union_exists(self):
+        qs1 = Number.objects.filter(num__gte=5)
+        qs2 = Number.objects.filter(num__lte=5)
+        with CaptureQueriesContext(connection) as captured_queries:
+            self.assertTrue(qs1.union(qs2).exists())
+        # union exists worked, but did it include too many columns?
+        self.assertEqual(len(captured_queries), 1)
+        qstr = captured_queries[0]['sql']
+        id, other_num = connection.ops.quote_name('id'), connection.ops.quote_name('other_num')
+        self.assertNotIn(id, qstr)
+        self.assertNotIn(other_num, qstr)
+        self.assertEqual(qstr.count("LIMIT 1"), 1)
+
+    @skipUnlessDBFeature('supports_limiting_in_compound')
+    def test_union_exists_limit_compound(self):
+        qs1 = Number.objects.filter(num__gte=5)
+        qs2 = Number.objects.filter(num__lte=5)
+        with CaptureQueriesContext(connection) as captured_queries:
+            self.assertTrue(qs1.union(qs2).exists())
+        # union exists worked, but did it include too many columns?
+        self.assertEqual(len(captured_queries), 1)
+        qstr = captured_queries[0]['sql']
+        id, other_num = connection.ops.quote_name('id'), connection.ops.quote_name('other_num')
+        self.assertNotIn(id, qstr)
+        self.assertNotIn(other_num, qstr)
+        self.assertEqual(qstr.count("LIMIT 1"), 3)
```
I guess these could be merged by doing a 

```python
self.assertEqual(
    qstr.count("LIMIT 1"),
    3 if connection.features.supports_limiting_in_compound else 1
)
```

charettes reviewed (approved):
Awesome work David, thanks for the adjustments.

I left comments regarding `connections[DEFAULT_DB_ALIAS]` and two tests that could be merged but otherwise the patch looks solid!

felixxm reviewed (commented):
@David-Wobrock Thanks for this patch :+1:

David-Wobrock commented:
Awesome @charettes 👍 that's exactly what I thought. I tried to pass the DB alias, because I didn't want to use the default one (because it's not necessarily the correct one), but I stumbled upon the call in `Exists.__init__` which didn't have this information when being called. Thanks a lot for the suggested improvement 🚀 
And also good call for merging the two test methods :)

Thanks for pointing this out @felixxm 👌  The patch became much more redable by using the existing DB feature.

I amended the second commit with the improvements

felixxm reviewed (commented):
@David-Wobrock Thanks for updates :+1: 

I pushed edits.

felixxm commented:
buildbot, test on oracle.
